[Woo POS] Bump minimum required WC version to 9.6, replace local products filters with downloadable parameter, page size 25

jaclync opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #1483
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This pull request updates products and variations and filtering criteria based on the latest decision pdfdoF-62l-p2#comment-7224. It removes the virtual local filter (as we want to show virtual products), replaces the downloadable local filter with a parameter in the API request that works for WC versions 9.6+, and updates the page size from 100 to 25 for products and variations requests.

Regarding the page size, I tested with the smallest font size on the tallest iOS device available - iPad Pro 13 in' so far in iOS 18.2. The minimum amount of items to fill the scroll view in portrait is 10. Even so, I picked 25 to match Android. Please feel free to share any thoughts on the page size.


<img src="https://github.com/user-attachments/assets/7e87f141-cbe9-4b93-b50a-197a8ea7479e" width="300" />

🗒️ It's likely we will update the minimum WC version requirement to 9.6 for POS based on p1736485809856869/1736484154.214279-slack-C070SJRA8DP. Since 9.6 is still in beta 2 at the time of writing, we will postpone the version requirement bump until the version is publicly released.

### Pagination and Product Parameters:

* [`Networking/Networking/Remote/ProductVariationsRemote.swift`](diffhunk://#diff-fa44db8cd4f05911d0bf515e02c8108729a7a649f31820a43ec91aee7306df67L331-R331): Changed the `variationsPerPage` constant from 100 to 25.
* [`Networking/Networking/Remote/ProductsRemote.swift`](diffhunk://#diff-dde78446a8994905cf951ddc694e4b3620a81a6dfd9e719f48afec2cd4c3e7b6R219): Added the `downloadable` parameter to the request and updated the `productsPerPage` constant from 100 to 25. [[1]](diffhunk://#diff-dde78446a8994905cf951ddc694e4b3620a81a6dfd9e719f48afec2cd4c3e7b6R219) [[2]](diffhunk://#diff-dde78446a8994905cf951ddc694e4b3620a81a6dfd9e719f48afec2cd4c3e7b6L651-R653) [[3]](diffhunk://#diff-dde78446a8994905cf951ddc694e4b3620a81a6dfd9e719f48afec2cd4c3e7b6R641)

### Test Updates:

* [`Networking/NetworkingTests/Remote/ProductsRemoteTests.swift`](diffhunk://#diff-01c6287fe6f095c7bdd598502a7d5a4d5b7c42b9291b20a415e3a242d2691dafR984-R996): Added a new test to verify that the `downloadable` parameter is set correctly in the `loadProductsForPointOfSale` method.

### Filtering Criteria:

* [`Yosemite/Yosemite/PointOfSale/PointOfSaleItemService.swift`](diffhunk://#diff-4ff6f858a82d655da0ffbd0c0620178056e0a60bd6a43fea67a3aafcf46219c9L58-L59): Removed the `isNotVirtual` and `isNotDownloadable` filtering functions and criteria. [[1]](diffhunk://#diff-4ff6f858a82d655da0ffbd0c0620178056e0a60bd6a43fea67a3aafcf46219c9L58-L59) [[2]](diffhunk://#diff-4ff6f858a82d655da0ffbd0c0620178056e0a60bd6a43fea67a3aafcf46219c9L130-L137)

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

- Switch to a store eligible for POS, has at least one **variable** and **downloadable** product, and has **WC version >= 9.6**
- Go to Menu > Point of Sale mode --> the downloadable product should not be included, and all simple products in the list should have a price
- Tap on a variable product --> variations should be loaded shortly, up to 25 per page

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

- [x] @jaclync tests that the products are loaded without the downloadable filter for stores on WC version < 9.6

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

Notice that the downloadable products are filtered out.

before (with local filters removed) | after
-- | --
![Simulator Screenshot - iPad Pro 11-inch (M4) - 2025-01-10 at 12 50 46](https://github.com/user-attachments/assets/43d7c050-f50b-402f-9ac9-0e05cf8d7217) | ![Simulator Screenshot - iPad Pro 11-inch (M4) - 2025-01-10 at 13 50 15](https://github.com/user-attachments/assets/13958cc0-a802-4ec5-9c08-3f83de3036db)



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

jaclync committed:
Remove local virtual and downloadable filters in favor of WC 9.6+ `virtual` and `downloadable` API support.

jaclync committed:
Remove `virtual` parameter from the latest decision pdfdoF-62l-p2#comment-7224.

jaclync committed:
Update products/variations page size from 100 to 25 for POS.

jaclync committed:
Nit: remove unused parameter.

jaclync committed:
Add a test case for the POS products request parameters.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

jaclync committed:
Remove downloadable from mock products API response. Update test case to include virtual product in the API response.

staskus commented on Yosemite/Yosemite/PointOfSale/PointOfSaleItemService.swift:
```diff
@@ -127,14 +125,6 @@ private extension PointOfSaleItemService {
-        !product.downloadable
-    }
-
     func hasPrice(product: Product) -> Bool {
```
Based on pdfdoF-62l-p2#comment-7285, this one can be removed as well 👍

jaclync replied:
Removed in https://github.com/woocommerce/woocommerce-ios/pull/14843/commits/b192a0c61d75395f810fe667293c93cc40b3906a.

staskus reviewed (approved):
This works as described. However, I think based on pdfdoF-62l-p2#comment-7285 it could be also expanded or done in another PR:

- Allow no-price products
- Restrict POS entry from 9.6+

jaclync commented:
Potential latest decision: p1736504688523299/1736484154.214279-slack-C070SJRA8DP

> * Empty-price and free products should be shown in POS
> * Virtual products should be shown in POS
> * Let's limit it to version 9.6+, but we need to ensure that it won't get released unless WooCommerce 9.6 is publicly available

Based on the proposed decision above, I will wait for WC 9.6 to be released this week before making all the changes:

- Remove empty price filter
- Update minimum WC version to 9.6

dangermattic commented:
<!--
  1 Error: This PR is tagged with <code>s...
  0 Warnings
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">This PR is tagged with <code>status: do not merge</code> label(s).</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

jaclync committed:
Merge branch 'trunk' into feat/14834-replace-local-filtering

jaclync committed:
Remove local price filter.

jaclync committed:
Bump minimum WC version from 6.6 to 9.6.

jaclync commented:
WC 9.6 is still not publicly available as of today. As I'm AFK after this Friday, I'm trying to get the PR ready for merge first and it can be merged when 9.6 is in GA.

@staskus it's not urgent, this PR is ready for review again with the changes in https://github.com/woocommerce/woocommerce-ios/pull/14843#issuecomment-2586025489.

jaclync commented:
Unit tests are failing, please skip the review for now and I'll update when it's ready for review again.

jaclync committed:
Update test case to include product without price now that the price filter is removed.

staskus commented:
> Unit tests are failing, please skip the review for now and I'll update when it's ready for review again.

TTP education tests should hopefully succeed if you merge now with trunk.

joshheald committed:
Merge branch 'trunk' into feat/14834-replace-local-filtering

jaclync commented:
Ready for review now! A test case failing from the removal of the price filter was updated in https://github.com/woocommerce/woocommerce-ios/pull/14843/commits/8916a21e1abc22ed064b5b5171a91713e993b332.

staskus reviewed (approved):
- POS only available from 9.6
- Downloadable filtered
- No price products shown

We'll need to implement additional handling for these no price or zero price products to finish the order.

joshheald commented:
@staskus thanks for the review; I think we're waiting to merge until we're sure 9.6 is released.

> We'll need to implement additional handling for these no price or zero price products to finish the order.

We discussed this last week – I think all that's needed is:

1. Don't start a card payment if the order total is 0.
2. Ensure cash payments feature flag is enabled before variable products.

In future, we could optimise by adding a `Mark order complete` button on free orders, rather than requiring merchants to open the cash flow.

Context: p1736742009959369-slack-C070SJRA8DP

iamgabrielma commented:
> 2\. Ensure cash payments feature flag is enabled before variable products.

👀  Let me know if I'm holding this one by having the flag disabled, it can be [enabled right away](https://github.com/woocommerce/woocommerce-ios/pull/14884) as soon as we merge [this PR](https://github.com/woocommerce/woocommerce-ios/pull/14913), as the rest it's just minor UI details.

joshheald committed:
Merge branch 'trunk' into feat/14834-replace-local-filtering

joshheald committed:
Merge branch 'trunk' into feat/14834-replace-local-filtering

joshheald committed:
14834 Only apply new WC version req when flag true

To retain the ability to turn off the feature flag for variable products, we should only use the higher minimum WC version when the flag is enabled.

If we disable the flag, downloadable products are still filtered out by local filters, and it’s no longer necessary to remove virtual products.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/235">21.5</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
14834 Update minimum version to include betas
